proposals/2023/03-15_init-step.md

```diff
@@ -0,0 +1,341 @@
+- `pipeline.Container.InitSteps`: an `InitStepSlice` / a slice of `pipeline.InitStep` structs that report on step/service init
+
+Future enhancement: also add `Mimetype` to the `Log` structs.
+
+### Server
+
+This is description is based on this draft PR: https://github.com/go-vela/server/pull/779
+
+#### Server API
+
+Add `api/initstep` package with endpoints for `InitStep` that mirror the `Step` endpoints:
+
+```
+- UpdateInitStep:     PUT    /api/v1/admin/initstep
+- CreateInitStep:     POST   /api/v1/repos/:org/:repo/builds/:build/initsteps
+- ListInitSteps:      GET    /api/v1/repos/:org/:repo/builds/:build/initsteps
+- GetInitStep:        GET    /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep
+- UpdateInitStep:     PUT    /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep
+- DeleteInitStep:     DELETE /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep
+- CreateInitStepLog:  POST   /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep/logs
+- GetInitStepLog:     GET    /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep/logs
+- UpdateInitStepLog:  PUT    /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep/logs
+- DeleteInitStepLog:  DELETE /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep/logs
+- PostInitStepStream: POST   /api/v1/repos/:org/:repo/builds/:build/initsteps/:initstep/stream
+```
+
+Add `middleware/initstep` package to inject `InitStep` in the `*/initsteps/*` APIs.
+
+Also include the endpoints in the mock server.
+
+#### Server Database
+
+Add `database/initstep` package for storing the new `InitStep` object:
+
+- `CountInitSteps`: gets the count of all InitSteps
+- `CountInitStepsForBuild`: gets the count of InitSteps by build ID
+- `CreateInitStep`: creates a new InitStep.
+- `DeleteInitStep`: deletes an existing InitStep.
+- `GetInitStep`: gets a InitStep by ID.
+- `GetInitStepForBuildgets`: gets an InitStep by build ID and InitStep number.
+- `ListInitSteps`: gets a list of all InitSteps.
+- `ListInitStepsForBuild`: gets a list of InitSteps by build ID.
+- `UpdateInitStep`: updates an existing InitStep.
+
+And methods for the InitStep Logs:
+- GetLogForInitStep: gets a log by init step ID from the database.
+
+#### End-User visible logging in the server
+
+Today, only the worker can safely add details to the init step log. With this change,
+the server can also add end-user visible logging, especially in places like the compiler
+where it would be helpful to highlight compile errors in the UI/CLI.
+
+In the Build and Webhook API endpoints, we can create the `InitStep` and `Log` as soon as we
+have the `RepoID` and `Build.Number`. Failures before this point can't be associated
+with a particular build, so they cannot bubble up to the end-user. The endpiont will
+then pass in a `library.Log` to the compiler that it creates specifically for the compiler.
+It can also record relevant its own log messages when handling requests like:
```

Comment:
Today, we don't create a build unless the pipeline compiles successfully.

Am I correct in understanding that we'd always generate a build now even if the pipeline doesn't compile?

Reply:
No. That is not my intention. The init step logs should only be created if the build has been created. I thought the build number could only be assigned once the build was persisted. Most of the compiler logic happens before the build is created, but not all of it (unless I misunderstand the sources).